File: src/jsp/solver/bruteforce.py
from itertools import permutations
from jsp.solver import JSPSolver
from jsp.domain import Jsp
from jsp.domain.solution import Schedule
from jsp.domain.solution import ScheduleItem

class BruteForceSolver(JSPSolver):
    def solve(self, jsp: Jsp) -> Schedule | None:
        if not jsp.jobs:
            return None

        max_machine_id = max(
            op.machine for job in jsp.jobs for op in job.operations
        )

        base_sequence: list[int] = []
        for idx, job in enumerate(jsp.jobs):
            base_sequence.extend([idx] * len(job.operations))

        best_makespan = float("inf")
        best_sequence: tuple[int, ...] | None = None
        seen_sequences: set[tuple[int, ...]] = set()

        for seq in permutations(base_sequence):
            if seq in seen_sequences:
                continue

            seen_sequences.add(seq)
            makespan = self._evaluate_makespan(jsp, seq, max_machine_id)

            if makespan < best_makespan:
                best_makespan = makespan
                best_sequence = seq

        if best_sequence is None:
            return None

        return self._build_schedule(jsp, best_sequence, max_machine_id)

    def _evaluate_makespan(
        self, jsp: Jsp, sequence: tuple[int, ...], max_machine_id: int
    ) -> int:
        job_end_time = [0] * jsp.total_jobs
        machine_end_time = [0] * (max_machine_id + 1)
        job_op_index = [0] * jsp.total_jobs

        for job_idx in sequence:
            job = jsp.jobs[job_idx]
            op_idx = job_op_index[job_idx]

            operation = job.operations[op_idx]
            m_id = operation.machine
            duration = operation.duration

            start_time = max(job_end_time[job_idx], machine_end_time[m_id])
            end_time = start_time + duration

            job_end_time[job_idx] = end_time
            machine_end_time[m_id] = end_time
            job_op_index[job_idx] += 1

        return max(job_end_time)

    def _build_schedule(
        self, jsp: Jsp, sequence: tuple[int, ...], max_machine_id: int
    ) -> Schedule:
        job_end_time = [0] * jsp.total_jobs
        machine_end_time = [0] * (max_machine_id + 1)
        job_op_index = [0] * jsp.total_jobs

        items: list[ScheduleItem] = []

        for job_idx in sequence:
            job = jsp.jobs[job_idx]
            op_idx = job_op_index[job_idx]

            operation = job.operations[op_idx]
            m_id = operation.machine
            duration = operation.duration

            start_time = max(job_end_time[job_idx], machine_end_time[m_id])
            end_time = start_time + duration

            job_end_time[job_idx] = end_time
            machine_end_time[m_id] = end_time
            job_op_index[job_idx] += 1

            items.append(
                ScheduleItem(
                    job=job.id,
                    op=op_idx + 1,
                    machine=m_id,
                    start=start_time,
                    end=end_time,
                    duration=duration,
                )
            )

        return Schedule(makespan=max(job_end_time), items=tuple(items))

File: src/solver/bruteforce.rs
use super::Solver;
use crate::domain::Jsp;
use crate::domain::solution::{Schedule, ScheduleItem};
use std::cmp;

pub struct BruteForce;

impl BruteForce {
    pub fn new() -> Self {
        Self
    }

    fn evaluate_makespan(&self, jsp: &Jsp, sequence: &[usize], max_machine_id: usize) -> usize {
        let mut job_end_time = vec![0; jsp.n_jobs()];
        let mut machine_end_time = vec![0; max_machine_id + 1];
        let mut job_op_index = vec![0; jsp.n_jobs()];

        for &job_idx in sequence {
            let job = &jsp.jobs()[job_idx];
            let op_idx = job_op_index[job_idx];

            let operation = &job.operations()[op_idx];
            let m_id = operation.machine();
            let duration = operation.duration();

            let start_time = cmp::max(job_end_time[job_idx], machine_end_time[m_id]);
            let end_time = start_time + duration;

            job_end_time[job_idx] = end_time;
            machine_end_time[m_id] = end_time;
            job_op_index[job_idx] += 1;
        }

        job_end_time.into_iter().max().unwrap_or(0)
    }

    fn build_schedule(&self, jsp: &Jsp, sequence: &[usize], max_machine_id: usize) -> Schedule {
        let mut job_end_time = vec![0; jsp.n_jobs()];
        let mut machine_end_time = vec![0; max_machine_id + 1];
        let mut job_op_index = vec![0; jsp.n_jobs()];

        let mut items = Vec::with_capacity(sequence.len());

        for &job_idx in sequence {
            let job = &jsp.jobs()[job_idx];
            let op_idx = job_op_index[job_idx];

            let operation = &job.operations()[op_idx];
            let m_id = operation.machine();
            let duration = operation.duration();

            let start_time = cmp::max(job_end_time[job_idx], machine_end_time[m_id]);
            let end_time = start_time + duration;

            job_end_time[job_idx] = end_time;
            machine_end_time[m_id] = end_time;
            job_op_index[job_idx] += 1;

            items.push(ScheduleItem::new(
                job.id(),
                op_idx + 1,
                m_id,
                start_time,
                end_time,
                duration,
            ));
        }

        let makespan = job_end_time.into_iter().max().unwrap_or(0);
        Schedule::new(makespan, items)
    }
}

impl Solver for BruteForce {
    fn solve(&self, jsp: &Jsp) -> Option<Schedule> {
        if jsp.jobs().is_empty() {
            return None;
        }

        let max_machine_id = jsp
            .jobs()
            .iter()
            .flat_map(|job| job.operations().iter())
            .map(|op| op.machine())
            .max()?;

        let mut base_sequence = Vec::new();
        for (idx, job) in jsp.jobs().iter().enumerate() {
            for _ in 0..job.operations().len() {
                base_sequence.push(idx);
            }
        }

        let mut best_makespan = usize::MAX;
        let mut best_sequence: Option<Vec<usize>> = None;

        let mut current_seq = base_sequence;

        if !current_seq.is_empty() {
            loop {
                let makespan = self.evaluate_makespan(jsp, &current_seq, max_machine_id);

                if makespan < best_makespan {
                    best_makespan = makespan;
                    best_sequence = Some(current_seq.clone());
                }

                if !next_permutation(&mut current_seq) {
                    break;
                }
            }
        }

        let best_seq = best_sequence?;
        Some(self.build_schedule(jsp, &best_seq, max_machine_id))
    }
}

fn next_permutation<T: Ord>(arr: &mut [T]) -> bool {
    if arr.len() <= 1 {
        return false;
    }
    let mut i = arr.len() - 1;
    while i > 0 && arr[i - 1] >= arr[i] {
        i -= 1;
    }
    if i == 0 {
        return false;
    }
    let mut j = arr.len() - 1;
    while arr[j] <= arr[i - 1] {
        j -= 1;
    }
    arr.swap(i - 1, j);
    arr[i..].reverse();
    true
}
